pkg/apis/config/feature_flags.go

```
@@ -105,9 +111,25 @@ func NewFeatureFlagsFromMap(cfgMap map[string]string) (*FeatureFlags, error) {
 	if err := setFeature(enableCustomTasks, DefaultEnableCustomTasks, &tc.EnableCustomTasks); err != nil {
 		return nil, err
 	}
+	setEnabledAPIFields(cfgMap, DefaultEnableAPIFields, &tc.EnableAPIFields)
 	return &tc, nil
 }
 
+// setEnabledAPIFields sets the "enable-api-fields" flag based on the content of a given map.
+// If the feature gate is invalid or missing then the flag is set to its default.
+func setEnabledAPIFields(cfgMap map[string]string, defaultValue string, feature *string) {
+	value := defaultValue
+	if cfg, ok := cfgMap[enableAPIFields]; ok {
+		value = strings.ToLower(cfg)
+	}
+	switch value {
+	case AlphaAPIFields, StableAPIFields:
+		*feature = value
+	default:
+		*feature = defaultValue
```

I don't have a strong opinion here (I like the way its implemented) but we are ignoring any value other than `stable` and `alpha` and defaulting to `stable` which is different from `bool` feature flags.  I think we should at least log this somewhere that its being defaulted to `stable`. On the other side, we could be ignoring a spelling mistake when someone tried to set it to `alpha` and by default we are operating `stable` APIs.

---

Right, I think it makes sense to at least log something in case of the value not being one we support, *but* the user will have to check the logs of the controller or the webhook (and know what to look for) to see it. But agreed we should log it anyway.

---

Indeed, I didn't want our machinery to end up in a broken state because the user mis-typed a value in the configmap.  But adding a log line is a great idea; I'll do that.

---

Hm.  We don't have access to `context` in a lot of places that leverage this code.  This might be a bigger change than I thought.

---

@pritidesai @vdemeester would you be ok with me using the go `log` package to print this message? The reason is that `NewFeatureFlagsFromConfigMap` is [passed into knative and called by the framework](https://github.com/tektoncd/pipeline/blob/879ba4f65732808a0614d76d5993af0bac736009/pkg/apis/config/store.go#L83) without a `ctx` variable. I can't think of a way to shim support for the logger into that function without completely rewriting how the config package is structured.  Since that function then [calls `NewFeatureFlagsFromMap`](https://github.com/tektoncd/pipeline/blob/879ba4f65732808a0614d76d5993af0bac736009/pkg/apis/config/feature_flags.go#L113), which in turn calls `setEnabledAPIFields` (from this PR), I don't see a great way to handle this except for leaning on the go `log` package.

Any other ideas on ways we could tackle this?

---

From the owners meeting just now:

1. We could return an error in a similar fashion as we do when boolean flags aren't correctly set to "true" / "false"
2. Knative pkg may be able to help us out here - they may have some configmap validation webhook tooling we can leverage to validate these on submission

---

I've gone ahead and implemented (1) - returning an error if the value is mis-spelled.  Created https://github.com/tektoncd/pipeline/issues/3895 to investigate validating the configmap contents somehow.